core/src/main/java/org/javarosa/core/model/instance/ExternalDataInstanceFactory.java

```diff
@@ -0,0 +1,51 @@
+package org.javarosa.core.model.instance;
+
+import org.javarosa.core.services.Logger;
+
+import java.util.Hashtable;
+
+/**
+ * Dispatchs on instance ID to create new external data instances.  Useful for
+ * injecting custom functionality for specific data instances, for instance if
+ * you want casedb template resolution to behave different than normal external
+ * instance resolution.
+ *
+ * @author Phillip Mates (pmates@dimagi.com)
+ */
+public class ExternalDataInstanceFactory {
+    private static final Hashtable<String, ExternalDataInstanceBuilder> instanceIdToBuilder =
```

Comment:
Hm, I would _really_ prefer to not increase the number of static builders/factories around the code. They are incredibly hard to sandbox and really harm the modularity of the code.

Can we make this a factory that we register with the static session or IIF or something? We already have an appealing entry point for when this would be injected.